### ISSUE TITLE
staticcheck: deferred error checks on wrong err

### ISSUE THREAD

**Issue by nicksnyder** [dominikh/go-tools#86]
Just ran into this problem in a codebase and thought it might be a good thing to add to staticcheck.

```
func incorrect() error {
	var err error
	defer func() {
		// Deferred error checks usually mean to check the returned function error,
		// not some other error in scope.
		if err != nil {
			fmt.Println(err)
		}
	}()
	if err := fmt.Errorf("will not be detected by defered func"); err != nil {
		return err
	}
	return nil
}

// Fix by naming the returned error.
func correct() (err error) {
	defer func() {
		if err != nil {
			fmt.Println(err)
		}
	}()
	if err := fmt.Errorf("will be detected by defered func"); err != nil {
		return err
	}
	return nil
}
```

**Comment by dominikh** [comment]
That'll be a difficult check to get right if we want to avoid false positives, since it is conceivable that people might write that code, but assign to `err` correctly.

Will have to look into some statistics to see if it's doable.

**Comment by shurcooL** [comment]
Can this be detected in SSA form as "checked `err` value is never assigned to, therefore this `if` statement is equivalent to `if false {`"?

**Comment by dominikh** [comment]
Not the way you're thinking, because `err` won't be a register, it will be a pointer, because of the way closures work. We can still determine that it is local and do what you suggest, but that will only capture a subset of the bugs the check should find. For example, this wouldn't be flagged, but really, absolutely should:

```
func fn() error {
	var err error
	defer func() {
		if err != nil {
			fmt.Println(err)
		}
	}()

	err = bar()
	if err != nil {
		return err
	}

	return errors.New("my custom error")
}
```

**Comment by nicksnyder** [comment]
> it is conceivable that people might write that code, but assign to err correctly.

True, but I would still flag the code as "suspicious" and the author could easily suppress the false positive by naming the return parameter instead.
